File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> keyToFreq = new HashMap<>();

        for(int it: nums){
            keyToFreq.put(it , keyToFreq.getOrDefault(it, 0) + 1);
        }

        Queue<Integer> heap = new PriorityQueue<>(
            (n1,n2) -> keyToFreq.get(n1) - keyToFreq.get(n2));


        for(int it: keyToFreq.keySet()){
            heap.add(it);
            if(heap.size() > k) heap.poll();
        }

        int[] ans = new int[k];
        for(int i=k-1;i>=0;i--){
            ans[i] = heap.poll();
        }

        return ans;
       
       
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;
        for(auto it: nums) map[it]++;

        priority_queue<pair<int,int>,
         vector<pair<int,int>>,
         greater<pair<int,int>>
         
         > pq; // by default max queue, convert to min

        for(auto [key, freq]: map){
            pq.push({freq,key});

            if(pq.size() > k) pq.pop();
        }
        vector<int> ans;
        while (!pq.empty()) {
            ans.push_back(pq.top().second);
            pq.pop();
        }
        return ans;
    }
};